lmdeploy/pytorch/weight_loader/model_weight_loader.py

```diff
@@ -0,0 +1,228 @@
+                model_path, weight_type)
+            self._load_shard(weight_path)
+            keys = list(self._state_dict.keys())
+            self._weight_map = dict((k, weight_name) for k in keys)
+
+        if adapters is None:
+            adapters = dict()
+
+        self._adapter_loaders: Dict[str, AdapterWeightLoader] = dict()
+        for ada_name, ada_path in adapters.items():
+            ada_loader = AdapterWeightLoader(ada_name, ada_path)
+            self._adapter_loaders[ada_name] = ada_loader
+
+    def _load_shard(self, path: str):
+        """load shards."""
+        self._state_dict.update(load_state_dict(path))
+
+    def _load_shard_for_key(self, key: str):
+        """load shard for key."""
+        if key in self._state_dict:
+            return
+        if key not in self._weight_map:
+            raise RuntimeError(f'Unknown weight: {key}.')
+
+        shard_file = osp.join(self.model_path, self._weight_map[key])
+        self._load_shard(shard_file)
+        if key not in self._state_dict:
+            raise RuntimeError(f'Can not found "{key}" in "{shard_file}"')
+
+    def pop(self, key: str):
+        """pop weight."""
+        key = self._prefix + key
+        self._load_shard_for_key(key)
+        return self._state_dict.pop(key)
+
+    def get(self, key: str):
+        """get weight."""
+        key = self._prefix + key
+        self._load_shard_for_key(key)
+        return self._state_dict.get(key)
+
+    def has(self, key: str):
+        """check has key."""
+        key = self._prefix + key
+        return key in self._weight_map
+
+    def adapter(self, key: str):
+        """get adapter loader."""
+        if key not in self._adapter_loaders:
+            raise RuntimeError(f'Unknown adapter: {key}')
+        return self._adapter_loaders[key]
+
+    @contextmanager
+    def prefix_context(self, mod_name: str):
+        """update prefix by mod name."""
+        old_prefix = self._prefix
+        if len(old_prefix) == 0:
+            new_prefix = f'{mod_name}.'
+        else:
+            new_prefix = f'{old_prefix}{mod_name}.'
+        self._prefix = new_prefix
+
+        with ExitStack() as stack:
+            for ada in self._adapter_loaders.values():
+                stack.enter_context(ada.prefix_context(mod_name))
+            yield new_prefix
+        self._prefix = old_prefix
+
+
+def _load_model_weights_impl(model: torch.nn.Module,
+                             loader: ModelWeightLoader,
+                             rank: int = 0,
+                             world_size: int = 1,
+                             device: torch.device = 'cpu'):
+    """load model weights implementation."""
+
+    def __load_no_recursive(mod: torch.nn.Module):
+        """load no recursive."""
+        for name, param in mod.named_parameters(recurse=False):
+            dtype = param.dtype
+            if not loader.has(name):
+                logger.debug(f'rank [{rank}]'
```

Comment:
How to invoke this condition?

Reply:
Some model might shared weight of token embedding, they do not safe redundant weight in checkpoint.